File: Stiver_Problems/DP/P017.cpp
// P017.cpp - Count subsets with difference k (S1 - S2 = k)

#include <bits/stdc++.h>
using namespace std;

// reduce to subset sum target = (k + totalSum)/2

long long countSubsetsTarget(const vector<int>& arr, int target) {
    int n = arr.size();
    vector<long long> prev(target + 1, 0), curr(target + 1, 0);
    prev[0] = 1;
    if (n > 0 && arr[0] <= target) prev[arr[0]]++;

    for (int i = 1; i < n; i++) {
        curr.assign(target + 1, 0);
        curr[0] = 1;
        for (int j = 1; j <= target; j++) {
            long long not_pick = prev[j];
            long long pick = 0;
            if (arr[i] <= j) pick = prev[j - arr[i]];
            curr[j] = pick + not_pick;
        }
        prev = curr;
    }
    return prev[target];
}

// brute
long long brute(const vector<int>& arr, int k) {
    int tot = accumulate(arr.begin(), arr.end(), 0);
    if ((k + tot) % 2 != 0) return 0;
    int target = (k + tot) / 2;
    if (target < 0) return 0;
    return countSubsetsTarget(arr, target);
}

// optimal is same formula, using DP
long long optimal(const vector<int>& arr, int k) {
    int tot = accumulate(arr.begin(), arr.end(), 0);
    if ((k + tot) % 2 != 0) return 0;
    int target = (k + tot) / 2;
    if (target < 0) return 0;
    return countSubsetsTarget(arr, target);
}

int main() {
    cout << "Enter arr: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> arr;
    int x;
    while (ss >> x) arr.push_back(x);
    cout << "Enter difference: ";
    int k;
    cin >> k;

    cout << "Brute: " << brute(arr, k) << "\n";
    cout << "Optimal: " << optimal(arr, k) << "\n";
    return 0;
}

File: DP/P017.py
# Let:
#     S1 + S2 = totalSum
# and
#     S1 − S2 = d

# Add both equations:

#     2S1 = d + totalSum
#     S1 = (d+totalSum)//2
# ✅ So we just need to count the number of subsets with sum = (d + totalSum)//2

def brute(arr,k):
    n=len(arr)
    tot=sum(arr)
    if (k+tot) % 2!=0:
        return 0
    target=(k+tot)//2
    def helper(i,target):
        if target==0:
            return 1
        if i==0:
            return 1 if target==arr[i] else 0
        not_pick=helper(i-1,target)
        pick=0
        if arr[i]<=target:
            pick=helper(i-1,target-arr[i])
        return pick+not_pick
    return helper(n-1,target)

def optimal(arr,k):
    n=len(arr)
    tot=sum(arr)
    if (k+tot) % 2!=0:
        return 0
    target=(k+tot)//2
    prev=[0]*(target+1)
    prev[0]=1
    if arr[0]<target:
        prev[arr[0]]=1
    for i in range(1,n):
        curr=[0]*(target+1)
        curr[0]=1
        for j in range(target+1):
            not_pick=prev[j]
            pick=0
            if arr[i]<=j:
                pick=prev[j-arr[i]]
            curr[j]=pick+not_pick
        prev=curr
    return prev[target]

arr=list(map(int,input("Enter arr: ").split()))
k=int(input("Enter difference: "))
print("Brute: ",brute(arr,k))
print("Optimal: ",optimal(arr,k))